rfcs/20180731-dockerfile-assembler.md

```diff
@@ -0,0 +1,164 @@
+Dockerfiles are duplicated, some are unused, and some users have made their own
+instead. None of the Dockerfiles use the ARG directive.
+
+Furthermore, `parameterized_docker_build.sh` is tightly coupled with the
+deploy-to-image-hub process we use, which is confusing because users who build
+the images locally don't need that information at all.
+
+This document proposes a new way for the TF team to maintain this complex set
+of similar Dockerfiles.
+
+# Design
+
+We use a generator to assemble multiple partial Dockerfiles into concrete
+Dockerfiles that get committed into source control. These Dockerfiles are fully
+documented and support argument customization. Unlike the parameterized image
+builder script, this system excludes the image deployment steps, which should
+be handled by a totally different system anyway.
+
+This section lightly describes the design, which is fully implemented in [this
+pull request to the main TensorFlow
+repo](https://github.com/tensorflow/tensorflow/pull/21291).
+
+Partial files are syntactically valid but incomplete files with Dockerfile
+syntax.
+
+Assembly is controlled by a specification file, defined in yaml. The spec
+defines the partials, the ARGs they use, the list of Dockerfiles to generate
+based on ordered lists of partials, and documentation for those values.
+
+The assembler is a python script that accepts a spec and generates a bunch of
+Dockerfiles to be committed. The spec includes documentation and descriptions,
+and the output Dockerfiles are fully documented and can be built manually.
+
+# Impact
+
+This approach has many convenient benefits:
+
+*   The result is concrete, buildable, documented Dockerfiles. Users who wish
+to build their own images locally do not need to also understand the build
+system.
+*   This implementation is agnostic to what images we would like to make
+available online (i.e. our Docker story). It's very easy to add new dockerfile
+outputs.
+*   The build-test-and-deploy-images process is decoupled from the Dockerfile
+generation process.
+*   Control of the set of dockerfiles is centralized to the spec file, instead
+of being spread across each Dockerfile.
+*   The spec can be extended to add more conveniences. My implementation, for
+example, already includes de-duplication of many similar Dockerfile
+specifications.
+*   All dockerfiles are consistently documented.
+*   Common pieces of code, like a slick shell environment or a Jupyter
+interface, can be updated in batch by updating a single partial file.
+*   The spec can also be used in the image building process, e.g. to read all
+available args.
+
+# Caveats and Rejected Alternatives
+
+I considered two alternatives while working on this.
+
+## Modern Multi-Stage Dockerfile
+
+"Multi-stage Building" is a powerful new Dockerfile feature that supports
+multiple FROM statements in one Dockerfile. It is meant to be used for creating
+artifacts with one image before using those artifacts in another image, but you
```

Comment:
I don't understand why you say "but" here. Is the intended meaning "in addition"?

Reply:
It's a mixture of both -- I revised this whole section, so it should be better now. Thanks!